src/main/java/com/mojang/serialization/codecs/BaseMapCodec.java

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+package com.mojang.serialization.codecs;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.mojang.datafixers.util.Pair;
+import com.mojang.datafixers.util.Unit;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.DynamicOps;
+import com.mojang.serialization.Lifecycle;
+import com.mojang.serialization.MapLike;
+import com.mojang.serialization.RecordBuilder;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+public interface BaseMapCodec<K, V> {
+
+    Codec<K> keyCodec();
+
+    Codec<V> elementCodec();
+
+    default <T> DataResult<Map<K, V>> decode(final DynamicOps<T> ops, final MapLike<T> input) {
+        final ImmutableMap.Builder<K, V> read = ImmutableMap.builder();
+        final ImmutableList.Builder<Pair<T, T>> failed = ImmutableList.builder();
+        final AtomicReference<DataResult<Unit>> result = new AtomicReference<>(DataResult.success(Unit.INSTANCE, Lifecycle.stable()));
```

Comment:
Don't use `AtomicReference` as generic mutable closure. Feels like that can be replaced with `Stream.reduce`

Reply:
I profiled array, AtomicReference, MutableObject and reduce, and there's no significant difference; replacing this with reduce since that's a bit cleaner, keeping AtomicReference in CompoundListCodec since that doesn't use streams (no reduce)

Reply:
(the rest of the code uses AtomicReference for stream refs so keeping that for consistency for now at least)